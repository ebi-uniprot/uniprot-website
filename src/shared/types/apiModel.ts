import { UniProtkbAPIModel } from '../../uniprotkb/adapters/uniProtkbConverter';
import { UniRefLiteAPIModel } from '../../uniref/adapters/uniRefConverter';
import { UniParcAPIModel } from '../../uniparc/adapters/uniParcConverter';
import { ProteomesAPIModel } from '../../proteomes/adapters/proteomesConverter';
import {
  TaxonomyAPIModel,
  Rank,
} from '../../supporting-data/taxonomy/adapters/taxonomyConverter';
import { KeywordsAPIModel } from '../../supporting-data/keywords/adapters/keywordsConverter';
import { CitationsAPIModel } from '../../supporting-data/citations/adapters/citationsConverter';
import { DiseasesAPIModel } from '../../supporting-data/diseases/adapters/diseasesConverter';
import { DatabaseAPIModel } from '../../supporting-data/database/adapters/databaseConverter';
import { LocationsAPIModel } from '../../supporting-data/locations/adapters/locationsConverter';

export type APIModel =
  | UniProtkbAPIModel
  | UniRefLiteAPIModel
  | UniParcAPIModel
  | ProteomesAPIModel
  | TaxonomyAPIModel
  | KeywordsAPIModel
  | CitationsAPIModel
  | DiseasesAPIModel
  | DatabaseAPIModel
  | LocationsAPIModel;

export type Statistics = {
  reviewedProteinCount: number;
  unreviewedProteinCount: number;
  // in citations
  computationallyMappedProteinCount?: number;
  communityMappedProteinCount?: number;
  // in taxonomy
  referenceProteomeCount?: number;
  proteomeCount?: number;
};

// TODO: remove all below and replace with types from corresponding namespaces
export type Lineage = {
  scientificName?: string;
  commonName?: string;
  synonyms?: string[];
  taxonId: number;
  rank: Rank;
  hidden?: boolean;
};

export type Xref = {
  database?: string;
  id?: string;
  properties?: { [key: string]: string };
  additionalIds?: string[];
  isoformId?: string;
  implicit?: true;
};
