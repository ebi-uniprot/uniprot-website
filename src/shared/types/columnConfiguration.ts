import { ReactNode } from 'react';

export type ColumRenderer<Schema extends Record<string, unknown>> = {
  label: ReactNode;
  tooltip?: ReactNode;
  // Exclude plain objects as it wouldn't detect an issue when returning
  // objects to React for rendering
  render: (data: Schema) => Exclude<ReactNode, Record<string, unknown>>;
};

export type ColumnConfiguration<
  Columns,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Schema extends Record<string, unknown> = any
> = Map<Columns, ColumRenderer<Schema>>;
