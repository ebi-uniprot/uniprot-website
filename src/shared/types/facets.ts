import { Facets as ARBAFacets } from '../../automatic-annotations/arba/config/ARBAFacetConfiguration';
import { Facets as UniRuleFacets } from '../../automatic-annotations/unirule/config/UniRuleFacetConfiguration';
import { Facets as ProteomesFacets } from '../../proteomes/config/ProteomesFacetConfiguration';
import { Facets as CitationsFacets } from '../../supporting-data/citations/config/CitationsFacetConfiguration';
import { Facets as DatabaseFacets } from '../../supporting-data/database/config/DatabaseFacetConfiguration';
import { Facets as DiseasesFacets } from '../../supporting-data/diseases/config/DiseasesFacetConfiguration';
import { Facets as KeywordsFacets } from '../../supporting-data/keywords/config/KeywordsFacetConfiguration';
import { Facets as LocationsFacets } from '../../supporting-data/locations/config/LocationsFacetConfiguration';
import { Facets as TaxonomyFacets } from '../../supporting-data/taxonomy/config/TaxonomyFacetConfiguration';
import { Facets as UniParcFacets } from '../../uniparc/config/UniParcFacetConfiguration';
import { Facets as UniProtKBFacets } from '../../uniprotkb/config/UniProtKBFacetConfiguration';
import { Facets as UniRefFacets } from '../../uniref/config/UniRefFacetConfiguration';

export type Facets =
  | UniProtKBFacets
  | UniRefFacets
  | UniParcFacets
  | ProteomesFacets
  | TaxonomyFacets
  | KeywordsFacets
  | CitationsFacets
  | DiseasesFacets
  | DatabaseFacets
  | LocationsFacets
  | UniRuleFacets
  | ARBAFacets;
