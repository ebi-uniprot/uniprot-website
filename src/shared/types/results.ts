import { ReactNode } from 'react';

export type FacetValue<Label extends ReactNode = string> = {
  // if from API, a simple string and always there,
  // but we need to make it an optional ReactNode for the constructed facets
  label?: Label;
  value: string;
  count: number;
};

export type FacetObject<Label extends ReactNode = string> = {
  // if from API, a simple string and always there,
  // but make it an optional React node for the constructed facets
  label?: Label;
  name: string;
  allowMultipleSelection: boolean;
  values?: Array<FacetValue<Label>>;
};

export type MatchedField = {
  name: string; // name of the field to use
  hits: number;
};

export type Suggestion = {
  query: string; // full suggested query
  hits: number; // Only an estimate!
};

export enum SearchResultsWarningCode {
  WildcardDetected = 41,
}

export type SearchResultsWarning = {
  code: SearchResultsWarningCode;
  message: string;
};

export type SearchResults<Schema> = {
  facets?: Array<FacetObject>;
  results: Array<Schema>;
  matchedFields?: Array<MatchedField>;
  suggestions?: Array<Suggestion>;
  warnings?: SearchResultsWarning[];
  obsoleteCount?: number;
  // Only if there's any async issue with the payload
  error?: string;
};

export type Response<Schema> = {
  data: SearchResults<Schema>;
  headers: {
    ['x-total-results']: string;
    link: string;
  };
};
