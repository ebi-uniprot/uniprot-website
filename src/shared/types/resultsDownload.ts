export enum ContentType {
  fasta = 'text/fasta',
  tsv = 'text/tsv',
  excel = 'application/vnd.ms-excel',
  xml = 'application/xml',
  rdfXml = 'application/rdf+xml',
  text = 'text/flatfile',
  gff = 'text/gff',
  list = 'text/list',
  json = 'application/json',
}
export enum FileFormat {
  fastaCanonical = 'FASTA (canonical)',
  fastaCanonicalIsoform = 'FASTA (canonical & isoform)',
  fastaRepresentative = 'FASTA (representative)',
  tsv = 'TSV',
  excel = 'Excel',
  xml = 'XML',
  rdfXml = 'RDF/XML',
  text = 'Text',
  gff = 'GFF',
  list = 'List',
  json = 'JSON',
}
