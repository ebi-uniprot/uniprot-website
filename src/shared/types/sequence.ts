export type Sequence = {
  value: string;
  length: number;
  molWeight: number;
  /**
   * @deprecated Should use MD5 instead
   */
  crc64: string;
  md5: string;
};
