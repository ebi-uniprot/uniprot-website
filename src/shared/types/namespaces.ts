import { JobTypes } from '../../tools/types/toolsJobTypes';

export enum Namespace {
  // Main data
  uniprotkb = 'uniprotkb',
  uniref = 'uniref',
  uniparc = 'uniparc',
  proteomes = 'proteomes',
  // Supporting data
  taxonomy = 'taxonomy',
  keywords = 'keywords',
  citations = 'citations',
  diseases = 'diseases',
  database = 'database',
  locations = 'locations',
  // Tools
  idmapping = 'id-mapping',
  // Annotations
  unirule = 'unirule',
  arba = 'arba',
  // UniSave
  unisave = 'unisave',
}

export const mainNamespaces = new Set<Namespace>([
  Namespace.uniprotkb,
  Namespace.uniref,
  Namespace.uniparc,
  Namespace.proteomes,
]);

export const supportingDataNamespaces = new Set<Namespace>([
  Namespace.taxonomy,
  Namespace.keywords,
  Namespace.citations,
  Namespace.diseases,
  Namespace.database,
  Namespace.locations,
]);

export const supportingDataAndAANamespaces = new Set<Namespace>([
  Namespace.taxonomy,
  Namespace.keywords,
  Namespace.citations,
  Namespace.diseases,
  Namespace.database,
  Namespace.locations,
  Namespace.unirule,
  Namespace.arba,
]);

export type SearchableNamespace = Exclude<
  Namespace,
  Namespace.idmapping | Namespace.unisave
>;

export const searchableNamespaceLabels: Record<SearchableNamespace, string> = {
  // Main data
  [Namespace.uniprotkb]: 'UniProtKB',
  [Namespace.uniref]: 'UniRef',
  [Namespace.uniparc]: 'UniParc',
  [Namespace.proteomes]: 'Proteomes',
  // Supporting data
  [Namespace.taxonomy]: 'Taxonomy',
  [Namespace.keywords]: 'Keywords',
  [Namespace.citations]: 'Literature citations',
  [Namespace.diseases]: 'Human diseases',
  [Namespace.database]: 'Cross-referenced databases',
  [Namespace.locations]: 'Subcellular locations',
  // Annotations
  [Namespace.unirule]: 'UniRule',
  [Namespace.arba]: 'ARBA',
};
export const toolResults = 'toolResults' as const;
export type ToolResults = typeof toolResults;
export type Searchspace = SearchableNamespace | ToolResults;
export const searchspaceLabels = {
  [toolResults]: 'Tool results',
  ...searchableNamespaceLabels,
};

export const namespaceAndToolsLabels: Record<Namespace | JobTypes, string> = {
  ...searchableNamespaceLabels,
  // Non-searchable namespace
  [Namespace.idmapping]: 'ID mapping',
  [Namespace.unisave]: 'UniProtKB entry history', // unused at the moment
  // Tools
  [JobTypes.ID_MAPPING]: 'Retrieve/ID mapping',
  [JobTypes.ALIGN]: 'Align',
  [JobTypes.ASYNC_DOWNLOAD]: 'File generation',
  [JobTypes.BLAST]: 'BLAST',
  [JobTypes.PEPTIDE_SEARCH]: 'Peptide search',
  [JobTypes.FOLDSEEK]: 'Foldseek',
};
