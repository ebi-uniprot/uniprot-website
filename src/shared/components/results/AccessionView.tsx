import { Link } from 'react-router-dom';
import cn from 'classnames';

import BasketStatus from '../../../basket/BasketStatus';
import EntryTypeIcon, { EntryType } from '../entry/EntryTypeIcon';

import { getEntryPath } from '../../../app/config/urls';

import { reIds } from '../../../tools/utils/urls';

import { Namespace, SearchableNamespace } from '../../types/namespaces';

import helper from '../../styles/helper.module.scss';
import styles from './styles/accession-view.module.scss';

type Props = {
  id: string;
  namespace: SearchableNamespace;
  entryType?: string | EntryType;
};

const AccessionView = ({ id, namespace, entryType }: Props) => {
  const { id: acc, start, end } = id.match(reIds)?.groups || { id };

  return (
    <span className={cn(helper['no-wrap'], styles['accession-view'])}>
      {entryType && <EntryTypeIcon entryType={entryType} />}
      <Link
        to={getEntryPath(
          namespace,
          acc,
          namespace === Namespace.uniprotkb || namespace === Namespace.uniparc
            ? 'entry'
            : undefined
        )}
        className={styles.accession}
      >
        {acc}
      </Link>
      {start && end && (
        <span>
          [{start}-{end}]
        </span>
      )}
      <BasketStatus id={id} />
    </span>
  );
};

export default AccessionView;
