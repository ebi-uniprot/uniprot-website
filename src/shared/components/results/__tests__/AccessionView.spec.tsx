import { screen } from '@testing-library/react';
import { Namespace } from '../../../types/namespaces';

import renderWithRouter from '../../../__test-helpers__/RenderWithRouter';

import AccessionView from '../AccessionView';

describe('AccessionView', () => {
  it('should have icon when proteome type is reference', () => {
    renderWithRouter(
      <AccessionView
        id={'UPI'}
        entryType="Reference proteome"
        namespace={Namespace.proteomes}
      />
    );
    expect(screen.queryByTitle(/reference Proteome entry/)).toBeInTheDocument();
  });

  it('should not have icon when proteome type is not reference', () => {
    renderWithRouter(
      <AccessionView
        id={'UPI'}
        entryType="Other proteome"
        namespace={Namespace.proteomes}
      />
    );
    expect(
      screen.queryByTitle(/reference Proteome entry/)
    ).not.toBeInTheDocument();
  });
});
