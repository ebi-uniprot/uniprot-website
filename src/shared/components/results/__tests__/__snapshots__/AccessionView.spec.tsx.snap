// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessionView should link only the accession leaving out the subset 1`] = `
<DocumentFragment>
  <span
    class="no-wrap accession-view"
  >
    <a
      class="accession"
      href="/uniprotkb/P59594/entry"
    >
      P59594
    </a>
    <span>
      [424-494]
    </span>
  </span>
</DocumentFragment>
`;
